[processing] Extent handling improvements

nyalldawson opened the pull request:
This PR:
- implements a new algorithm to allow feature extraction (with optional clipping) to an extent. We don't currently have an easy way to pull features out of a layer (or clip a layer) to a rectangular extent, so that's addressed by this algorithm.
- Uses QgsReferencedRectangles for extent parameters, so that CRS handling and transparent reprojection is done when required. This allows selection of an extent on the canvas in say EPSG:4326, but is used to clip a layer in EPSG:3785.

nyalldawson committed:
[FEATURE] New processing algorithm "extract/clip by extent"

Allows extract a subset of another layer using an extent, with
optional setting to clip geometries to the extent

nyalldawson committed:
Allow use of Qgs(Referenced)Rectangle for processing extent parameter values

And add a new target CRS argument to parameterAsExtent. If set, and
the source CRS of the rectangle parameter can be determined, then
the returned value will be the rectangle automatically reprojected
to the desired target CRS.

nyalldawson committed:
Nicer display of rectangle parameter values

nyalldawson committed:
Return a referenced rectangle from ExtentSelectionPanel

nyalldawson committed:
Use CRS when calculating extent during algorithm extection

nyalldawson committed:
Use geometry of reproject extent parameters for more accurate clipping

nyalldawson committed:
Fix passing QgsReferencedRec/Point from python to c++

nyalldawson commented:
Some notes:
- if acceptable, i'll do the same thing for point parameters
- i haven't updated the ported GDAL algorithms to handle transparent reprojection of extents... because I'm not familiar enough with those algorithms to know which crs the extent should be in. But this should be a couple of second change for someone who knows those algs better :)

nirvn commented:
Now that QGIS' extent widget has dragging onto canvas support, might be worth considering using it in processing. 

I love extract by extent, it'll be extremely useful within models.

nyalldawson committed:
Fix unit test data

nyalldawson committed:
Add method to recover extent parameter CRS

nyalldawson committed:
New algorithm to convert an extent parameter to a layer

Creates a new layer with a single feature with polygon geometry
covering the extent parameter value.

This is designed for use in models where some child algorithms require
a layer based input, while others require an extent based
parameter

alexbruy commented:
+1 to merge and update point parameters to use similar approach.

I'm working on GDAL algorithms and will try to address extent handling where possible.
